Add support for observing and modifying the selected text range.

theMomax opened the pull request:
This PR addresses #2.

 The selected text range can be observed and modified via another `Binding` now:
 
 ```swift
  struct ContentView: View {
     static private let initialSource = "let a = 42\n"

     @State private var source = Self.initialSource
     @State private var selection = Self.initialSource.endIndex..<Self.initialSource.endIndex

     var body: some View {
         CodeEditor(source: $source,
                    selection: $selection,
                    language: .swift,
                    theme: .ocean,
                    autoscroll: false)
         Button("Select All") {
             selection = source.startIndex..<source.endIndex
         }
     }
 }
 ```
 When `autoscroll` is `true`, the editor automatically scrolls to the respective
 cursor position when `selection` is modfied from the outside, i.e. programatically.

helje5 commented:
Thanks, I'll try to find the time to look at that tomorrow. Actually I implemented this feature myself a few weeks back, but didn't end up using it, because the performance was abysmal (e.g. moving up and down in even small texts became horribly slow due to all the SwiftUI refreshes attached to the selection changes). Does that work fine with the patch?

theMomax commented:
Thanks for taking a look! I didn't notice any performance impact. I only tried on macOS, but I didn't edit anything OS specific, so it should still work for UIKit devices.

theMomax commented:
Turns out it was a good idea to do some more tests...

I just fixed a bug where sometimes the cursor position resulting from user interaction would be overridden by the old value in the `selection` `Binding`. In this example I'm pasting the same code twice in rapid succession and sometimes the cursor gets stuck in between the two pasted sections:

https://user-images.githubusercontent.com/21169289/147088000-a4d46c13-59d5-44f8-8ada-333a9d21ded0.mov

Here you can see the fixed version and also some further interaction where at least I can't see any performance issues.

https://user-images.githubusercontent.com/21169289/147088028-a8cc58dd-5426-49ae-99ab-95dd5d3e8b26.mov

helje5 commented:
Sigh, this now conflicts with the other PR :-) Could you move the selection callback out of the if? Does it work on iOS and macOS, btw?

Otherwise this looks fine, I think I'm going to just merge it.

theMomax commented:
@helje5 I updated the PR. I resolved the merge conflict and added the iOS implementation. I tested on both platforms.

I also noticed during testing on iOS that programatically modifying the text source `Binding` can cause `State` updates during view rendering, which can cause undefined behavior. I thus added the same guard for the on-text-change callbacks as I used on the on-text-selection-change callbacks.

I think this is ready to merge now.
